File: products/src/application/useCases/CreateProductUseCase.ts
import { AbstractProductRepository } from "src/infra/repositories/AbstractProductRepository";
import { CreateProductDto } from "../../application/dtos/CreateProductDto";
import { ResponseProductDto } from "../../application/dtos/ResponseProductDto";
import { Product } from "src/domain/entities/Product";
import { Barcode } from "src/domain/valueObjects/barcode";

/**
 * Use case class responsible for creating a new product.
 */
export class CreateProductUseCase {
    
    private productRepository: AbstractProductRepository;

    /**
     * Constructor of the CreateProductUseCase class.
     * @param {AbstractProductRepository} productRepository - The repository for product operations.
     */
    constructor(productRepository: AbstractProductRepository){
        this.productRepository = productRepository; 
    }

    /**
     * Method to execute the creation of a new product.
     * @param {CreateProductDto} data - The data required to create the product.
     * @returns {Promise<ResponseProductDto>} - A promise that resolves to the created product.
     * @throws {Error} - If any required field is missing or if the code length exceeds 13 characters.
     */
    async execute(data: CreateProductDto): Promise<ResponseProductDto> {
        const entity = new Product(
            data.name,
            data.code,
            data.description,
            data.value
        )
        const productAlreadyExist = await this.productRepository.getByCode(entity.code);
        if (productAlreadyExist){
            throw new Error("The product's code already exists")
        }
        const product = await this.productRepository.create(data); 
        return product;
    }
}

File: cart-api/src/main/java/com/example/cartapi/application/useCases/product/CreateProductUseCase.java
/**
 * Use case implementation for creating a new product.
 * This class implements the CreateProductUseCaseInterface.
 * It handles the business logic for creating a product entity,
 * including validation and interaction with the product repository.
 */
package com.example.cartapi.application.useCases.product;

import com.example.cartapi.domain.entities.product.ProductEntityFactory;
import com.example.cartapi.domain.gateways.repositories.ProductRepositoryInterface;
import com.example.cartapi.domain.dtos.product.CreateProductRecordDTO;
import com.example.cartapi.domain.entities.product.ProductEntity;
import com.example.cartapi.domain.errors.ValidationException;
import com.example.cartapi.domain.gateways.useCases.product.CreateProductUseCaseInterface;

import java.util.UUID;

public class CreateProductUseCase implements CreateProductUseCaseInterface {
    private final ProductRepositoryInterface productRepositoryInterface;

    /**
     * Constructor for CreateProductUseCase.
     *
     * @param productRepositoryInterface The repository for accessing and managing product data.
     */
    public CreateProductUseCase(ProductRepositoryInterface productRepositoryInterface) {
        this.productRepositoryInterface = productRepositoryInterface;
    }

    /**
     * Executes the use case to create a new product.
     *
     * @param createProductRecordDTO The data transfer object containing information to create the product.
     * @throws ValidationException if the product already exists.
     */
    @Override
    public void execute(CreateProductRecordDTO createProductRecordDTO) throws ValidationException {
        ProductEntity productAlreadyExist = productRepositoryInterface.findByExternalID(createProductRecordDTO.externalID());
        if(productAlreadyExist == null){
            throw new ValidationException("Product already exist");
        }

        ProductEntity productEntity = ProductEntityFactory.createProductEntity(
                createProductRecordDTO.externalID(),
                createProductRecordDTO.code(),
                createProductRecordDTO.value()
        );

        productRepositoryInterface.create(productEntity);
    }
}
